crates/driver/src/infra/simulator/mod.rs

```diff
@@ -114,11 +129,46 @@ enum Inner {
 }
 
 #[derive(Debug, thiserror::Error)]
-pub enum Error {
+pub enum SimulatorError {
     #[error("tenderly error: {0:?}")]
     Tenderly(#[from] tenderly::Error),
     #[error("blockchain error: {0:?}")]
     Blockchain(#[from] blockchain::Error),
     #[error("enso error: {0:?}")]
     Enso(#[from] enso::Error),
 }
+
+#[derive(Debug, thiserror::Error)]
+#[error("err: {err}, tx: {tx:?}")]
+pub struct Error {
+    err: SimulatorError,
+    /// Optional since we are not interested in the tx data for http
+    /// communication errors, for example.
```

Comment:
This seems to indicate that we should have different error variants.
One for network related issues without `tx` and one for simulation reverts with `tx`.

Reply:
Agree that explicit error types will make the intention simpler to understand downstream. Otherwise we'd later on decide whether to call notify depending on whether some optional value is None or not (which is very implicit and requires domain knowledge from some other part of the codebase).

Reply:
Makes sense. fixed.

Reply:
Hmm, I had something else in mind.
Right now we have simulator errors (`Tenderly`, `BlockChain`, `Enso`). I think all of them have some error variants related to http errors and some related to the simulation reverting. What I had in mind was to have the new error type abstract those differences away and have 1 category for http errors and 1 for simulation errors and `with_tx` would pattern match on all those simulator error variants and categorize them into common categories. That way we can keep specific error categories without caring about whether they originated from `Tenderly`, the node or `Enso`.

The solution we have at the moment would for example allow you to turn an error originating from a revert into the new error kind but without forcing you to create the tx data for it.

So concretely I would suggest:
```rust
enum Error {
    Network(reqwest::Error),
    Revert{
        // revert message techincally not needed but might speed up debugging simple errors
        // also external solvers can compare the error they got with ours in case something doesn't
        // add up
        message: String,
        tx: eth::Tx,
    }
}
```

Alternatively we could only have the broad error categories and directly return them from the individual simulator implementations (this seems like the most correct solution to me).

Reply:
Seems nice but I'm not sure how to categorize blockchain error to only these 2 types. Would have to have a few more.

Reply:
Also simulator source (tenderly, blockchain, enso) seems like an important info to have for debugging to understand the results...

Reply:
The source depends on the config though and doesn't change at runtime (so it's pretty easy to find which type of estimator was used)

Reply:
Tried to follow the idea but did not end up with a good enough solution and also had ownership errors in the conversion method I couldn't end the fight with.
@MartinquaXD would you be willing to do improvement in the follow up PR as you like?